### PR TITLE
Make stack honor --skip flag

### PR DESCRIPTION
**Pull request by vanceism7**
This should fix the issue of libs not being skipped when specified in the skip flags #3827.
I added skipping for both internal libs and foreign libs, but not sure if this is totally correct or needed. Just let me know if I need to change anything, the fix was pretty minimal anyhow

### CODE REVIEW

File: test/integration/tests/3827-skip-libs-bug/Main.hs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import StackTest
+
+main :: IO ()
+main = do
+
+  -- The lib, exe, and test components of "this" package will fail if built.
```
[the remark this answers is not in the document]
Comment (vanceism7) [reply test/integration/tests/3827-skip-libs-bug/Main.hs]: Do you mean passing `"--resolver", defaultResolverArg` in the args array for `stack`?

File: test/integration/tests/3827-skip-libs-bug/Main.hs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import StackTest
+
+main :: IO ()
+main = do
```
[the remark this answers is not in the document]
Comment (vanceism7) [reply test/integration/tests/3827-skip-libs-bug/Main.hs]: Do you mean passing `"--resolver", defaultResolverArg` in the args array for `stack`?

**Comment by vanceism7** [comment]
I see what you mean about the failing build for the this lib. I'm not sure how the whole flags aspect of builds works. It looks like they are flags belonging to cabal build system itself, is there some sort of docs I can find for those? I tried looking but all I really found was cabal-install stuff. I could also alternatively just do a check to see if the `.stack-work` directory exists instead of making a failing build?

File: test/integration/tests/3827-skip-libs-bug/Main.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbaynard) [reply test/integration/tests/3827-skip-libs-bug/Main.hs]: Yes, but you don't need the `"--resolver"` bit — `defaultResolverArg` should do.

**Comment by dbaynard** [comment]
Yes, they are Cabal flags. I'll take a look at this.

**Comment by dbaynard** [comment]
Thanks for your patience, @vanceism7. I've done the basic setup; I'm just running into other bugs while trying to choose the actual test commands!

Perhaps ominously, `stack build --skip this:lib this:lib` doesn't seem to work, even on your version, so it's a little hard for me to judge. Would you please confirm which test cases you can get to work (and which you can't)?

File: test/integration/tests/3827-skip-libs-bug/Main.hs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  -- stack should try to build the tests and fail due to missing tests flag
+  failsToBuild targetTest [buildLib]
+
+  -- stack should try to build the testcutable and succeed
+  builds targetTest [buildLib, buildTest]
+
+  stack [defaultResolverArg, "clean", "--full"]
+
+withArgs :: ([String] -> IO()) -> String -> [[String]] -> IO()
+withArgs stackFunction target argsList = stackFunction . mconcat $ [defaultResolverArg, "build"] : argsList <> [[target]]
```
Comment (dbaynard) [inline test/integration/tests/3827-skip-libs-bug/Main.hs]:
```suggestion
withArgs stackFunction target argsList = stackFunction . mconcat $ [[defaultResolverArg, "build"], argsList, [target]]
```

File: test/integration/tests/3827-skip-libs-bug/Main.hs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+builds, failsToBuild :: String -> [[String]] -> IO()
+builds = withArgs stack
+failsToBuild = withArgs stackErr
+
+skipLib, skipExe, skipTest, buildLib, buildExe, buildTest :: [String]
+skipLib   = ["--skip", "this:lib"]
+skipExe   = ["--skip", "this:test"]
```
Comment (dbaynard) [inline test/integration/tests/3827-skip-libs-bug/Main.hs]:
```suggestion
skipExe   = ["--skip", "this:exe"]
```

File: test/integration/tests/3827-skip-libs-bug/Main.hs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+skipTest  = ["--skip", "this:exe"]
```
Comment (dbaynard) [inline test/integration/tests/3827-skip-libs-bug/Main.hs]:
```suggestion
skipTest  = ["--skip", "this:test"]
```

**Comment by snoyberg** [comment]
@vanceism7 Just pinging on open PRs, are you still planning on continuing with this?

**Comment by vanceism7** [comment]
Hey sorry guys, I've meant to respond but have been a little busy. I believe my original tests were working but they've been changed pretty drastically I think, and I don't know that my tests were really the correct tests to use in the first place. I don't think I'll really have time to finish up the testing issues for this though. Sorry about that!

**Comment by vanceism7** [comment]
Hey sorry guys, I've meant to respond but have been a little busy. I believe my original tests were working but they've been changed pretty drastically I think, and I don't know that my tests were really the correct tests to use in the first place. I don't think I'll really have time to finish up the testing issues for this though. Sorry about that!

**Comment by vanceism7** [comment]
Hey sorry guys, I've meant to respond but have been a little busy. I believe my original tests were working but they've been changed pretty drastically I think, and I don't know that my tests were really the correct tests to use in the first place. I don't think I'll really have time to finish up the testing issues for this though. Sorry about that!

**Comment by snoyberg** [comment]
Understood @vanceism7, no problem. I'll try to get this over the finish line, thanks for your work on this PR 👍

**Comment by snoyberg** [comment]
Alright, after digging into this a bit more, it's going to make a lot more sense to try and tackle this _after_ we have component based builds in place, so I'm going to delay this for now. CC @qrilka

**Comment by snoyberg** [comment]
The ultimate solution for this problem will likely be quite different, so I'm going to close this out for now.

**Pull request closed without merging**